recipe/README.md

```diff
@@ -98,3 +98,86 @@ of 2024-11-28:
 | sympy     | ==1.13.1       | >=1.13.1, !=1.13.2 | 1.13.3 | (wheel metadata)             |
 | typing-extensions | >=4.8.0 |       | 4.12.2      | (wheel metadata)                    |
 | triton    | 3.1.0          | none   | 3.1.0       | (wheel metadata)                    |
+
+
+Maintenance notes
+=================
+
+Packages built by the recipe
+----------------------------
+The recipe currently builds four packages:
+
+1. `libtorch` that installs the common libraries, executables and data files
+   that are independent of selected Python version and are therefore shared
+   by all Python versions.
+
+2. `libtorch-cuda-linalg` that provides the shared `libtorch_cuda_linalg.so`
```

Comment:
is this the preffered one?

as in, is it preffered that users make use of linalg?

I ask because we try very hard to ensure that

```
mamba install pytorch
```
installs the best hardware optimized one.

Reply:
Not sure I understand the question.

`libtorch_cuda_linalg.so` is required for some operations. It is always pulled in by `libtorch` itself, i.e. the end result is that some version of the library is always installed.

As for magma vs. nomagma, I think we ought to prefer magma. https://github.com/conda-forge/pytorch-cpu-feedstock/issues/275#issue-2588203757 suggests that cusolver will be faster for other workflows, but the magma build supports both and according to https://pytorch.org/docs/stable/backends.html#torch.backends.cuda.preferred_linalg_library, it has a heuristic to choose the faster backend for given operation.

Reply:
So who benefits from making this optional? 

Sorry if this obvious.

Reply:
This is #298, i.e. people who want to avoid installing the large magma dependency (~250M in package). Given that `libtorch-cuda-linalg` is 200k (nomagma) / 245k (magma), I've figured out it's worth it.

Reply:
The no-magma variant reduces the on-disk install size of a `pytorch-gpu` install from 7 GB to 5 GB (as #275 says). So this is definitely worth doing. 

Given that `cusolver` is now on average fairly close in performance to `magma` _and_ that the PyTorch wheels don't include magma, I'd argue that the default should be `nomagma` (I'd certainly almost always want that for my use cases). However, what the default is is less important than both options being available.

Reply:
I'd also explain the size difference explicitly in the README, so it's clear what the tradeoff is.

Reply:
ok this makes sense, thanks for the in depth discussion.

I am weary of "following upstream". So instead of "this is what upstream does on pypi", I want to make sure we are asking

1. Do we believe that magma usage is niche enough to omit by default and should be an opt in.

it seems that @mgorny thinks we should make it opt out, while @rgommers thinks we should make it opt in. I'll let you all decide, but hopefully you can document your intentions.

Over the years, we've just gone through many different ideas (track features, build numbers, etc....) to help users get a good default experience.

Reply:
I don't really have a strong opinion on it. My suggestion was basically "minimal change", but I'm fine with either default.

Reply:
> Pytorch statically links to magma. 

Argh, I missed that, hard to tell from an unpacked wheel. 

So is the 2 GB size difference explained by the Magma shared libraries being that large but PyTorch only needing a small fraction of the functionality in Magma, or is there more to it?

Reply:
Well, I don't know about the latter point, but yes, indeed, `libmagma` and `libmagma_sparse` take 2G:

```
$ du -hs $CONDA_PREFIX
7.4G	/home/mgorny/.conda/envs/pytorch
$ conda uninstall --force libmagma libmagma_sparse

## Package Plan ##

  environment location: /home/mgorny/.conda/envs/pytorch

  removed specs:
    - libmagma
    - libmagma_sparse


The following packages will be REMOVED:

  libmagma-2.8.0-h0af6554_0
  libmagma_sparse-2.8.0-h0af6554_0


Proceed ([y]/n)? y


Downloading and Extracting Packages:

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
$ du -hs $CONDA_PREFIX
5.4G	/home/mgorny/.conda/envs/pytorch
```

Reply:
`libmagma` itself is [around](https://anaconda.org/conda-forge/libmagma/files) 250MB, and `libmagma_sparse` [<10MB](https://anaconda.org/conda-forge/libmagma_sparse/files). So the vast majority of that footprint comes through transitive dependencies (which might also be depended on by something else in the environment, erasing the gains).

The `libmagma` [build](https://conda-metadata-app.streamlit.app/?q=conda-forge%2Flinux-64%2Flibmagma-2.8.0-hfdb99dd_0.conda) for CUDA 11.8 depends on `cudatoolkit`, which is pretty hefty, but then, the CUDA 11.8 builds are fast becoming obsolete. The CUDA 12.0+ [build](https://conda-metadata-app.streamlit.app/?q=conda-forge%2Flinux-64%2Flibmagma-2.8.0-h0af6554_0.conda) mainly depends on `libcublas` (~260MB) and `libcusparse` (~120MB).

The rest of the dependencies are either tiny or ubiquitous (libblas, libgcc, libstdcxx), so I don't think the 2GB footprint is realistic. It sounds more like ~650MB to me. Which is still big (so no strong opinion whether we want it by default), but two thirds less than 2GB.

Reply:
My numbers were actually for installed artifacts, i.e. literally:

```
-rwxrwxr-x 2 mgorny mgorny 2.0G Jul 22 22:55 /home/mgorny/.conda/envs/pytorch/lib/libmagma.so
```

Reply:
> My numbers were actually for installed artifacts, i.e. literally:

Ok, thanks. Better compression ratio than I expected! Perhaps as an additional data point, could we compare the total unpacked footprint for a pytorch environment with/without magma?

2GB is huge in absolute terms, but perhaps less so if we're talking relatively (say if it's ~10-15%).

I'd do it myself but currently travelling and on a metered connection, so too big to try right now

Reply:
Thanks. That certainly supports the argument that it should be opt-in (I guess as another alternative we could consider packaging static libmagma and using that here, if purely to follow the upstream default while not wasting gratuitous space on disk and the wire - but that's not an obvious win either).

Reply:
I don't understand why it is obvious that it should be opt-in.

To me, nvidia and fat binaries go together. trimming away magma is an optimization.

I think that ultimately we should prefer that for runtime features, users "opt-out" of them, once advanced users have learned their workflows, rather than the surprise of new users being surprised that we are missing a feature.

Reply:
> I don't understand why it is obvious that it should be opt-in.

To me the argument is (even though I generally prefer the maximalist approach to packaging!) that 2GB for a feature that few users will touch is a pretty gratuitous waste of bandwith/storage, and those who need it (assuming their percentage among users is reasonably small) should opt-in.

IIUC, you raised a similar [point](https://github.com/conda-forge/ffmpeg-feedstock/pull/279#issuecomment-2526217286) in ffmpeg (about an extra 450MB). That said, I only wrote "supports the argument that it should be opt-in" - I don't feel strongly enough about either option to insist on something.

Reply:
i understsand. it is just not "obvious" thats all.

Reply:
>  few users will touch

Is this based on your experience or facts?

The main driving principle with the pytorch package here has been to provide the same features as in upstream conda/wheels so that users who do use our packages are not driven away by mysterious performance issues. With the deprecation of anaconda.org/pytorch channel, we certainly do not want the users of that channel to look at conda-forge and form the opinion that the pytorch packages in conda-forgeare sub-par. Therefore I do strongly prefer using the same options as the upstream wheels which in this case is with magma.

Reply:
> Is this based on your experience or facts?

It's based on @rgommers' statement "10 or so functions, vs. [...] all of libtorch". I don't know pytorch nearly as well as Ralf, so I caveated my point with "assuming their percentage [of those who need it] among users is reasonably small".

IMO it's always a numbers game (and often we don't know the numbers, so we stay conservative; which is fine!) - if we knew, say, that <0.1% of users are affected, then burdening 99.9% users with a useless 2GB download sounds unappealing. The question gets progressively harder the higher we estimate the percentage. Probably anything north of 5% of users, we'd default to performance rather than footprint? I don't have a magic bullet. 🤷 

> provide the same features as in upstream conda/wheels so that users who do use our packages are not driven away by mysterious performance issues.

I agree with the overall direction of course. We still have a few things to catch up (e.g. cpuarch builds, as IIUC pytorch effectively [requires](https://github.com/pytorch/pytorch/blob/v2.5.1/cmake/MiscCheck.cmake#L42-L48) something between x86_64 v3 and v4).

That said, it's certainly technically possible to warn users of a potential performance-cliff where we don't yet support certain upstream optimizations (or require an opt-in for whatever reason). It's mainly a question of how much of that we want to patch and/or upstream.

To be clear, I'm also OK with "fat binaries are preferable to performance degradation or extra maintenance effort". The magma-case is just not so black-and-white IMO. What do you think about the idea of making an exception to link libmagma statically?

Reply:
To be fair, they're ~10 pretty important linear algebra functions. The default matching upstream is also a good argument. I actually wish that was done more often, so I can't complain about the argument being applied here:) For example, cuDNN version is very performance-sensitive, so I wish that were matched better rather than just taking whatever the latest version is.

Static linking or some other way to reduce `libmagma` in size would be good to investigate then though, because binary size is also a performance (and usability) issue.

Reply:
They are pretty important linear algebra functions like lu_solve, so I would say vast majority of users would use magma. Not the other way. For eg: a simple script like below
```
import torch
a = torch.randn(5, 5).cuda()
b = torch.randn(5, 5).cuda()
torch.linalg.solve(a, b)
```
would invoke magma. The number of functions provided is a poor indicator of its usage. (but certainly valid about code size)

> e.g. cpuarch builds, as IIUC pytorch effectively [requires](https://github.com/pytorch/pytorch/blob/v2.5.1/cmake/MiscCheck.cmake#L42-L48) something between x86_64 v3 and v4).

pytorch requires those from the compiler, but not at runtime. fbgemm does runtime cpu dispatching just like openblas, numpy, etc.

> What do you think about the idea of making an exception to link libmagma statically?

Why do you think that linking libmagma statically will help? Just because upstream used a static build? One option that's different from upstream is https://github.com/conda-forge/libmagma-feedstock/pull/22

Reply:
Somehow missed the last ping here. Thanks for the info, you certainly know pytorch much better than I do.

> Why do you think that linking libmagma statically will help?

Presumably those 10 pytorch functions do not make use of the entirety of the 2GB libmagma. If they were compiled against a static build, the impact would be much smaller. 

> One option that's different from upstream is [conda-forge/libmagma-feedstock#22](https://github.com/conda-forge/libmagma-feedstock/pull/22)

That's even better though! :)

Reply:
Including this bit from that merged PR thread:

Me:
> Did you measure the difference between the two approaches? Would be interested to know how they compare

Isuru:
> 2GB -> 620 MB

xref: https://github.com/conda-forge/libmagma-feedstock/pull/22#issuecomment-2533448783

<hr>

IOW a nearly 70% reduction in size. Perhaps it is worth rechecking the analysis given this change? The biggest thing might be something else now

Reply:
It looks even better on my machine after @isuruf's compression fix:

```
$ mamba create -n pytorch-gpu-19dec2024 pytorch-gpu
$ cd ~/mambaforge/envs/pytorch-gpu-19dec2024/
$ du -hsx * | sort -rh | head -3
4,1G    lib
1,6G    targets
69M     include

$ du -hsx lib/* | sort -rh | head -10
1,4G    lib/libtorch_cuda.so
432M    lib/libcudnn_engines_precompiled.so.9.3.0
296M    lib/python3.13
293M    lib/libmagma.so
276M    lib/libtorch_cpu.so
252M    lib/libnccl.so.2.23.4
233M    lib/libcudnn_adv.so.9.3.0
104M    lib/libcudnn_ops.so.9.3.0
91M     lib/libmagma_sparse.so
68M     lib/libmkl_core.so.2

$ du -hsx targets/x86_64-linux/lib/* | sort -rh | head -10
469M    targets/x86_64-linux/lib/libcublasLt.so.12.6.4.1
280M    targets/x86_64-linux/lib/libcusparse.so.12.5.4.2
266M    targets/x86_64-linux/lib/libcufft.so.11.3.0.4
146M    targets/x86_64-linux/lib/libcusolver.so.11.7.1.2
104M    targets/x86_64-linux/lib/libcublas.so.12.6.4.1
92M     targets/x86_64-linux/lib/libcurand.so.10.3.7.77
86M     targets/x86_64-linux/lib/libcusolverMg.so.11.7.1.2
56M     targets/x86_64-linux/lib/libnvrtc.so.12.6.85
49M     targets/x86_64-linux/lib/libnvJitLink.so.12.6.85
5,1M    targets/x86_64-linux/lib/libnvrtc-builtins.so.12.6.85
```

So 385 MB left for `libmagma.so` + `libmagma_sparse.so`.

I'll also note that there are two `linux-64/libmagma` [packages](https://anaconda.org/conda-forge/libmagma/files), and the CUDA 12 one seems to be a lot smaller than the CUDA 11 one. There are no interpretable build strings, but comparing package hashes with the logs from https://github.com/conda-forge/libmagma-feedstock/pull/22 shows that CUDA 11 is the larger one. The analysis above uses the `h7847c38_1` package.

![image](https://github.com/user-attachments/assets/e4c8fa73-4e91-4c34-ae7c-4e85f9d4dcc4)

Reply:
Well, I'm not opposed to reverting the magma parts. Just let me know if I should do it.

Reply:
> Well, I'm not opposed to reverting the magma parts. Just let me know if I should do it.

For completeness here: Isuru moved this decision conversation to gh-275, which seems like the right thing to do for visibility.